src/Workspaces/Remote/Core/SolutionAssetProvider.cs

```diff
@@ -29,20 +35,66 @@ public SolutionAssetProvider(HostWorkspaceServices services)
+
+            if (checksums.Length == 1)
             {
-                using var writer = new ObjectWriter(outputStream, leaveOpen: false, cancellationToken);
+                singleAsset = (await assetStorage.GetAssetAsync(scopeId, checksums[0], cancellationToken).ConfigureAwait(false)) ?? SolutionAsset.Null;
+            }
+            else
+            {
+                assetMap = await assetStorage.GetAssetsAsync(scopeId, checksums, cancellationToken).ConfigureAwait(false);
+            }
+
+            // Work around the lack of async stream writing in ObjectWriter, which is required when writing to the RPC pipe.
+            // Run two tasks - the first synchronously writes to a local pipe and the second asynchronosly transfers the data to the RPC pipe.
+            //
+            // Configure the pipe to never block on write (waiting for the reader to read). This prevents deadlocks but might result in more
+            // (non-contiguous) memory allocated for the underlying buffers. The amount of memory is bounded by the total size of the serialized assets.
+            var localPipe = new Pipe(RemoteHostAssetSerialization.PipeOptionsWithUnlimitedWriterBuffer);
 
-                var assetStorage = _services.GetRequiredService<ISolutionAssetStorageProvider>().AssetStorage;
-                var serializer = _services.GetRequiredService<ISerializerService>();
-                await RemoteHostAssetSerialization.WriteDataAsync(writer, assetStorage, serializer, scopeId, checksums, cancellationToken).ConfigureAwait(false);
-            }, cancellationToken);
+            Task.Run(() =>
+            {
+                try
+                {
+                    using var stream = localPipe.Writer.AsStream(leaveOpen: false);
+                    using var writer = new ObjectWriter(stream, leaveOpen: false, cancellationToken);
+                    RemoteHostAssetSerialization.WriteData(writer, singleAsset, assetMap, serializer, scopeId, checksums, cancellationToken);
+                }
+                catch (Exception e) when (FatalError.ReportWithoutCrashUnlessCanceled(e, cancellationToken))
+                {
+                    // no-op
+                }
+            }, cancellationToken).Forget();
+
+            async Task CopyPipeData()
+            {
+                Exception? exception = null;
+                try
+                {
+                    await localPipe.Reader.CopyToAsync(pipeWriter, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception e)
+                {
+                    FatalError.ReportWithoutCrashUnlessCanceled(e, cancellationToken);
+                    exception = e;
+                }
+                finally
+                {
+                    await localPipe.Reader.CompleteAsync(exception).ConfigureAwait(false);
+                    await pipeWriter.CompleteAsync(exception).ConfigureAwait(false);
```

Comment:
what does `CompleteAsync(Exception)` mean for pipes?

Reply:
It marks the operation as completed with failure.

Reply:
If a reader reads from a pipe and the writer completes, the reader knows there is no more data coming.

Reply:
ah i see.  ok.  so reader/writer is not just an one-and-done stream. it's something that can be continually written to or read from. and you can (presumably in a non-blocking fashion) wait on either other end for the other side.  but there's an explicit way to signal things are done, which the other side will see and which will end teh waits, and cause no more data to be readable/writable.

Seems a lot like nodejs streams in that way.  Nice that we have somethign like this.